pallets/dynamic-evm-base-fee/src/lib.rs

```diff
@@ -0,0 +1,234 @@
+// This file is part of Astar.
+
+// Copyright (C) 2019-2023 Stake Technologies Pte.Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+// Astar is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Astar is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Astar. If not, see <http://www.gnu.org/licenses/>.
+
+//! Dynamic Evm Base Fee Pallet
+//!
+//! ## Overview
+//!
+//! The pallet is responsible for calculating `Base Fee Per Gas` value, according to the current system parameters.
+//! This is not like `EIP-1559`, instead it's intended for `Astar` and `Astar-like` networks, which allow both
+//! **Substrate native transactions** (which in `Astar` case reuse Polkadot transaction fee approach)
+//! and **EVM transactions** (which use `Base Fee Per Gas`).
+//!
+//! For a more detailed description, reader is advised to refer to Astar Network forum post about [Tokenomics 2.0](https://forum.astar.network/t/astar-tokenomics-2-0-a-dynamically-adjusted-inflation/4924).
+//!
+//! ## Approach
+//!
+//! The core formula this pallet tries to satisfy is:
+//!
+//!     base_fee_per_gas = adjustment_factor * weight_factor * 25 / 98974
+//!
+//! Where:
+//! * **adjustment_factor** - is a value that changes in-between the blocks, related to the block fill ratio.
+//! * **weight_factor** - fixed constant, used to convert consumed _weight_ to _fee_.
+//!
+//! The implementation doesn't make any hard requirements on these values, and only requires that a type implementing `Get<_>` provides them.
+//!
+//! ## Implementation
+//!
+//! The core logic is implemented in `on_finalize` hook, which is called at the end of each block.
+//! This pallet's hook should be called AFTER whicever pallet's hook is responsible for updating **adjustment factor**.
+//!
+//! The hook will calculate the ideal new `base_fee_per_gas` value, and then clamp it in between the allowed limits.
+//!
+//! ## Interface
+//!
+//! Pallet provides an implementation of `FeeCalculator` trait. This makes it usable directly in `pallet-evm`.
+//!
+//! A _root-only_ extrinsic is provided to allow setting the `base_fee_per_gas` value manually.
+//!
+//! ## Practical Remarks
+//!
+//! According to the proposed **Tokenomics 2.0**, max amount that adjustment factor will be able to change on live networks in-between blocks is:
+//!
+//! adjustment_new = adjustment_old * (1 + adj + adj^2/2)
+//!
+//! adj = v * (s - s*)
```

Comment:
I read the report but just wanted to confirm to make sure. It would be helpful if you can add a comment to suggest these values are from report right next to those values (I know it's already on the top).

Reply:
I'll do that.